src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -109,6 +117,96 @@ public void write(String str, int off, int len) throws IOException {
         }  
     }
 
+    private class UnfinishedLine {
+        final long threadId;
+        int curCapacity = 512;
+        byte[] buffer = new byte[curCapacity];
+        int curEnd = 0;
+        int lastNewLine = 0;
+
+        public UnfinishedLine() {
+            this.threadId = Thread.currentThread().getId();
+        }
+
+        /**
+         * Adding input combines previously unfinished sentences with possible
+         * new sentences. A number of cases come together here that otherwise
+         * should be diligently separated. 
+         * 
+         *   - An unfinished line can already exist or not
+         *   - The new input can contain newlines or not
+         *   - The new input can end with a newline character or not
+         * 
+         * By concatenating the previous unfinished line with the new input
+         * all we have to do now is figure out where the last newline character is.
+         * 
+         */
+        private void add(byte[] newInput, int offset, int len) {
+            // first ensure capacity of the array
+            if (curEnd + len >= curCapacity) {
+                var oldCapacity = curCapacity;
+                curCapacity *= 2; // this should not happen to often. we're talking a few lines of text.
```

Comment:
maybe we should grow by 50%?

Reply:
the chance of having to grow again is large if we only grow by 50%. This way we quickly grow and we don't grow again on short notice.